src/accumulator/proof.rs

```diff
@@ -241,6 +243,293 @@ impl Proof {
 
         Ok((nodes, calculated_root_hashes))
     }
+    /// Uses the data passed in to update a proof, creating a valid proof for a given
+    /// set of targets, after an update. This is useful for caching UTXOs. You grab a proof
+    /// for it once and then keep updating it every block, yielding an always valid proof
+    /// over those UTXOs.
+    pub fn update(
+        self,
+        cached_hashes: Vec<sha256::Hash>,
+        add_hashes: Vec<sha256::Hash>,
+        block_targets: Vec<u64>,
+        remembers: Vec<u64>,
+        update_data: UpdateData,
+    ) -> Result<(Proof, Vec<sha256::Hash>), String> {
+        let (proof_after_deletion, cached_hashes) = self.update_proof_remove(
+            block_targets,
+            cached_hashes.clone(),
+            update_data.new_del,
+            update_data.prev_num_leaves,
+        )?;
+
+        let data_after_addition = proof_after_deletion.update_proof_add(
+            add_hashes,
+            cached_hashes.clone(),
+            remembers,
+            update_data.new_add,
+            update_data.prev_num_leaves,
+            update_data.to_destroy,
+        )?;
+
+        Ok(data_after_addition)
+    }
+    fn update_proof_add(
+        self,
+        adds: Vec<sha256::Hash>,
+        cached_del_hashes: Vec<sha256::Hash>,
+        remembers: Vec<u64>,
+        new_nodes: Vec<(u64, sha256::Hash)>,
+        before_num_leaves: u64,
+        to_destroy: Vec<u64>,
+    ) -> Result<(Proof, Vec<sha256::Hash>), String> {
+        // Combine the hashes with the targets.
+        let orig_targets_with_hash: Vec<(u64, sha256::Hash)> = self
+            .targets
+            .to_owned()
+            .into_iter()
+            .zip(cached_del_hashes.into_iter())
+            .collect();
+
+        // Attach positions to the proof.
+        let proof_pos = get_proof_positions(
+            &self.targets,
+            before_num_leaves,
+            util::tree_rows(before_num_leaves),
+        );
+        let proof_with_pos = proof_pos
+            .clone()
+            .into_iter()
+            .zip(self.hashes.clone())
+            .collect();
+
+        // Remap the positions if we moved up a after the addition row.
+        let targets_after_remap = Proof::maybe_remap(
+            before_num_leaves,
+            adds.len() as u64,
+            orig_targets_with_hash.clone(),
+        );
+        let mut final_targets = targets_after_remap.clone();
+        let mut new_nodes_iter = new_nodes.iter();
+        let proof_with_pos =
+            Proof::maybe_remap(before_num_leaves, adds.len() as u64, proof_with_pos);
+        // remembers is an index telling what newly created UTXO should be cached
+        for remember in remembers {
```

Comment:
Those hashes look a little weird, `Add 6 leaves [0, 1, 2, 3, 4, 5], then delete [4, 5] and add [6, 7, 8]` and `delHashes` is 
```
040000ff00000000000000000000000000000000000000000000000000000000
050000ff00000000000000000000000000000000000000000000000000000000
```
ff000004 and ff000005 in little-endian, and 31 bytes with 0's. I'm almost certain this is not their hashes.  

If you meant to use this hash as leaf hash anyway without knowing the preimage, this won't work with this testing framework we made because it hashes the preimages (usually numbers) and use this as leaf hash.

Reply:
Ah the 0xff's are shoved in there in the go tests because go library interprets alls 0s as nothing. So instead of hashing it, the test just picks a byte and turns it into 0xff.

I'll change the test so it's better suited for this lib. But the:
```
040000ff00000000000000000000000000000000000000000000000000000000
050000ff00000000000000000000000000000000000000000000000000000000
```
 is supposed to represent `4` and `5`.

Reply:
Could you add this into the tests? This is the same as above but it's hashing the ints.

```
    {
        "update": {
            "adds": [6, 7, 8],
            "proof": {
                "targets": [4, 5],
                "hashes": []
            },
            "del_hashes": [
                "e52d9c508c502347344d8c07ad91cbd6068afc75ff6292f062a09ca381c89e71",
                "e77b9a9ae9e30b0dbdb6f510a264ef9de781501d7b6b92ae89eb059c5ab743db"
            ]
        },
        "remembers": [1],
        "cached_proof": {
            "targets": [3],
            "hashes": [
                "dbc1b4c900ffe48d575b5da5c638040125f65db0fe3e24494b76ea986457d986",
                "02242b37d8e851f1e86f46790298c7097df06893d6226b7c1453c213e91717de"
            ]
        },
        "initial_stump": "Add 6 leaves [0, 1, 2, 3, 4, 5], then delete [4, 5] and add [6, 7, 8]",
        "initial_roots": [
            "df46b17be5f66f0750a4b3efa26d4679db170a72d41eb56c3e4ff75a58c65386",
            "9eec588c41d87b16b0ee226cb38da3864f9537632321d8be855a73d5616dcc73"
        ],
        "initial_leaves": 6,
        "cached_hashes": [
            "084fed08b978af4d7d196a7446a86b58009e636b611db16211b65a9aadff29c5"
        ],
        "expected_roots": [
            "0f1ff6dc2fc7cf8e27236d51de9af7d32adf10a94eef498e167d5e7cf98b0112",
            "beead77994cf573341ec17b58bbf7eb34d2711c993c1d976b128b3188dc1829a"
        ],
        "expected_targets": [3, 19],
        "expected_cached_hashes": [
            "084fed08b978af4d7d196a7446a86b58009e636b611db16211b65a9aadff29c5",
            "ca358758f6d27e6cf45272937977a748fd88391db679ceda7dc7bf1f005ee879"
        ]
    }
```